File: packages/xyd-opencli2go/src/command.ts
import type { Command } from '@xyd-js/opencli';

import { Imports, goBool, goFile, goSlice, goStr, goStruct, lit, type GoVal } from './golit';
import { renderFlags } from './flags';
import { renderHandler } from './handler';
import { buildLeafModel } from './model';
import { pascalCase, splitWords } from './naming';

const CLI = 'github.com/urfave/cli/v3';

function renderCommand(
  command: Command,
  pathNames: string[],
  module: string,
  imports: Imports,
  handlers: string[],
): GoVal {
  imports.add(CLI);
  const fields: [string, GoVal][] = [['Name', goStr(command.name)]];
  if (command.aliases?.length) fields.push(['Aliases', goSlice('string', command.aliases.map(goStr))]);
  if (command.description) fields.push(['Usage', goStr(command.description)]);
  if (command.hidden) fields.push(['Hidden', goBool(true)]);

  if (command.commands?.length) {
    const subs = command.commands.map((sub) =>
      renderCommand(sub, [...pathNames, sub.name], module, imports, handlers),
    );
    fields.push(['Commands', goSlice('*cli.Command', subs)]);
  } else if (command['x-openapi']) {
    const flags = renderFlags(buildLeafModel(command).flags);
    if (flags.length) fields.push(['Flags', goSlice('cli.Flag', flags)]);
    const handler = renderHandler(pathNames, command, module, imports);
    fields.push(['Action', lit(handler.name)]);
    handlers.push(handler.code);
  }

  return goStruct('cli.Command', fields, true);
}

export interface ResourceFile {
  path: string;
  content: string;
  constructor: string;
}

/** Render one `pkg/cmd/<resource>.go` file for a top-level command + its subtree. */
export function renderResourceFile(topCommand: Command, module: string): ResourceFile {
  const imports = new Imports();
  const handlers: string[] = [];
  const struct = renderCommand(topCommand, [topCommand.name], module, imports, handlers);

  const ctorName = `New${pascalCase(topCommand.name)}Command`;
  const ctor = `func ${ctorName}() *cli.Command {\n\treturn ${struct(1)}\n}`;
  const content = goFile('cmd', imports, [ctor, ...handlers]);

  const fileBase = splitWords(topCommand.name).join('') || 'command';
  return { path: `pkg/cmd/${fileBase}.go`, content, constructor: ctorName };
}

File: crates/xyd_opencli2go/src/command.rs
//! Command tree + resource-file rendering — port of command.ts.

use serde_json::Value;

use crate::flags::render_flags;
use crate::golit::{go_bool, go_file, go_slice, go_str, go_struct, lit, GoVal, Imports};
use crate::handler::render_handler;
use crate::model::build_leaf_model;
use crate::naming::{pascal_case, split_words};

const CLI: &str = "github.com/urfave/cli/v3";

fn render_command(
    command: &Value,
    path_names: &[String],
    module: &str,
    imports: &mut Imports,
    handlers: &mut Vec<String>,
) -> GoVal {
    imports.add(&[CLI]);
    let name = command.get("name").and_then(|n| n.as_str()).unwrap_or("");
    let mut fields: Vec<(String, GoVal)> = vec![("Name".to_string(), go_str(name))];

    if let Some(aliases) = command.get("aliases").and_then(|a| a.as_array()) {
        if !aliases.is_empty() {
            let rendered = aliases
                .iter()
                .filter_map(|a| a.as_str())
                .map(go_str)
                .collect();
            fields.push(("Aliases".to_string(), go_slice("string", rendered)));
        }
    }
    if let Some(desc) = command
        .get("description")
        .and_then(|d| d.as_str())
        .filter(|s| !s.is_empty())
    {
        fields.push(("Usage".to_string(), go_str(desc)));
    }
    if command.get("hidden") == Some(&Value::Bool(true)) {
        fields.push(("Hidden".to_string(), go_bool(true)));
    }

    let sub_commands = command.get("commands").and_then(|c| c.as_array());
    if let Some(subs) = sub_commands.filter(|s| !s.is_empty()) {
        let rendered: Vec<GoVal> = subs
            .iter()
            .map(|sub| {
                let sub_name = sub
                    .get("name")
                    .and_then(|n| n.as_str())
                    .unwrap_or("")
                    .to_string();
                let mut child_path = path_names.to_vec();
                child_path.push(sub_name);
                render_command(sub, &child_path, module, imports, handlers)
            })
            .collect();
        fields.push(("Commands".to_string(), go_slice("*cli.Command", rendered)));
    } else if command.get("x-openapi").is_some() {
        let model = build_leaf_model(command);
        let flags = render_flags(&model.flags);
        if !flags.is_empty() {
            fields.push(("Flags".to_string(), go_slice("cli.Flag", flags)));
        }
        let handler = render_handler(path_names, command, module, imports);
        fields.push(("Action".to_string(), lit(handler.name)));
        handlers.push(handler.code);
    }

    go_struct("cli.Command", fields, true)
}

pub struct ResourceFile {
    pub path: String,
    pub content: String,
    pub constructor: String,
}

/// Render one `pkg/cmd/<resource>.go` file for a top-level command + subtree.
pub fn render_resource_file(top_command: &Value, module: &str) -> ResourceFile {
    let mut imports = Imports::new();
    let mut handlers: Vec<String> = Vec::new();
    let top_name = top_command
        .get("name")
        .and_then(|n| n.as_str())
        .unwrap_or("")
        .to_string();
    let go_struct_val = render_command(
        top_command,
        std::slice::from_ref(&top_name),
        module,
        &mut imports,
        &mut handlers,
    );

    let ctor_name = format!("New{}Command", pascal_case(&top_name));
    let ctor = format!(
        "func {ctor_name}() *cli.Command {{\n\treturn {}\n}}",
        go_struct_val(1)
    );
    let mut decls = vec![ctor];
    decls.extend(handlers);
    let content = go_file("cmd", &imports, &decls);

    let file_base = {
        let joined = split_words(&top_name).join("");
        if joined.is_empty() {
            "command".to_string()
        } else {
            joined
        }
    };
    ResourceFile {
        path: format!("pkg/cmd/{file_base}.go"),
        content,
        constructor: ctor_name,
    }
}
